### DISCUSSION TITLE
STM32F405RGT6 servo motor help

### DISCUSSION THREAD

**Discussion by GoesWithoutSaying** [micropython/micropython#14138] [STM32 / Pyboard]
I'm new to pyboard, during the fist attempt to control sg90 servo motor, i found that the pulse_width is extremely large at pin X1 and it's different if i swith the timer, it's about 35000-216000 for the motor to move 180 degree. after some reserch, i understand that is Timer's prescaler and period issue.
so the question is: what to do to make pulse width normally like the SG90's doc tells(freq=50hz, pulse_width=0.5ms-2.5ms for 0-180),
control sg90 is only a small experiment, further i will try to control some big servo motors(to build a motion controller), what should i learn, any guiance?
here is the code
```
timer = pyb.Timer(2, freq=50)
ch1 = timer.channel(1, mode=pyb.Timer.PWM, pin=pyb.Pin("X1"))
ch1.pulse_width(100000)
pyb.delay(2000)
timer.deinit()
```

**Comment by GoesWithoutSaying** [comment]
new issue. 
usb_vcp.read() block evrything untill next data received

`usb_vcp = USB_VCP()

while True :
    if usb_vcp.any():
        data = usb_vcp.read()
        pyb.delay(1000)
        LED(3).toggle()`

LED(3)will keep light untill next data received, this happens often, but sometimes it is normal, LED(3) will toggle

**Comment by GitHubsSilverBullet** [comment]
First of all, A SG90 is **not a servo motor,** it's a servo.
Given that piece of information, why aren't you make use of the Servo class?

**Reply by GoesWithoutSaying** [reply]
thank you for reply. control SG90 now is not a problem, by using Timer class and setting pulse_width_percent, SG90 is working properly
here is the code
`tim2 = Timer(2, freq=50)
a = tim2.channel(1, mode=Timer.PWM, pin=Pin.board.X1)
a.pulse_width_percent(12.5)`
USB_VCP communication is now bothering me, the problem is during usb_vcp.read(), it seems that usb_vcp.read() blocked next code untill i send same message, otherwise, LED(4) aways keep light on, even i set pyb.delay(1000). i tried delete all code but USB_VCP part, it is still happens, i need the received data immediatly. sometimes it working as i expect, but not often.
